### PR TITLE
Implement dms

### PR DESCRIPTION
**Pull request by junha1**
Please check [this](https://github.com/junha1/serde-tc/blob/main/serde-tc/examples/calculator.rs)
for the use of the library `serde-tc`.

### CODE REVIEW

File: network/src/dms.rs [hunk 1]
```diff
@@ -67,6 +83,133 @@ trait DistributedMessageSetRpcInterface: Send + Sync + 'static {
...
+                        .get(&port_key)
+                        .ok_or_else(|| anyhow!("can't find port key: {}", port_key))?
+                ),
+                reqwest::Client::new(),
+            )));
+            let messages = stub
+                .get_message(height, known_messages_)
+                .await
+                .map_err(|e| anyhow!(e))?
+                .map_err(|e| anyhow!(e))?;
```
Comment (lemonholic) [inline network/src/dms.rs:194]:
```suggestion
                .await?
                .map_err(|e| anyhow!(e))?;
```
Let's use a shorter way.

File: network/src/dms.rs [hunk 2]
```diff
@@ -67,6 +83,133 @@ trait DistributedMessageSetRpcInterface: Send + Sync + 'static {
     ) -> Result<Vec<RawMessage>, String>;
 }
 
+#[async_trait]
+impl<S: Storage> DistributedMessageSetRpcInterface for StorageWrapper<S> {
+    async fn get_message(
+        &self,
+        height: BlockHeight,
+        knowns: Vec<Hash256>,
+    ) -> Result<Vec<RawMessage>, String> {
```
Comment (lemonholic) [inline network/src/dms.rs:92]: Why do you use String as the error instead of anyhow?

File: network/src/dms.rs [hunk 3]
```diff
@@ -80,35 +223,58 @@ trait DistributedMessageSetRpcInterface: Send + Sync + 'static {
...
+    /// The interval of the direct-peer fetch. If none, it will fetch only in `fetch()`, not in `serve()`.
+    pub fetch_interval: Option<Duration>,
+}
+
+impl<N: GossipNetwork, S: Storage> DistributedMessageSet<N, S> {
     /// Creates a new and empty storage with the given directory.
     /// If there is already a directory, it discards everything and creates a new one.
     /// You should try `open()` first!
     ///
     /// - `dms_key`: The unique key for distinguishing the DMS.
```
Comment (lemonholic) [inline network/src/dms.rs:245]:
```suggestion
    /// - `dms_key`: The key for distinguishing the type of the DMS instance. (e.g. governance, consensus, ...)
```
Suggest improved comment

File: network/src/dms.rs [hunk 4]
```diff
@@ -117,35 +283,137 @@ impl<N: P2PNetwork, S: Storage> DistributedMessageSet<N, S> {
...
+            State {
+                height: state.height + 1,
+                key: state.key,
+            },
+        )
+        .await?;
+        Ok(())
     }
 
     /// Serves the p2p network node and the RPC server indefinitely, constantly updating the storage.
```
Comment (lemonholic) [inline network/src/dms.rs:328]:
```suggestion
    /// Serves the gossip-network node and the RPC server indefinitely, constantly updating the storage.
```

File: network/src/dms.rs [hunk 5]
```diff
@@ -117,35 +283,137 @@ impl<N: P2PNetwork, S: Storage> DistributedMessageSet<N, S> {
...
+        let peers_ = peers.clone();
+        let network_config_ = network_config.clone();
+        let fetch_task = async move {
+            let interval = if let Some(x) = self.config.fetch_interval {
+                x
+            } else {
+                return Result::<(), Error>::Ok(());
+            };
+            loop {
+                let peers = peers_.read().await.to_vec();
```
Comment (lemonholic) [inline network/src/dms.rs:346]:
```suggestion
                let peers = peers_.read().await;
```
redundant `to_vec`

File: network/src/dms.rs [hunk 6]
```diff
@@ -117,35 +283,137 @@ impl<N: P2PNetwork, S: Storage> DistributedMessageSet<N, S> {
...
+        let storage_ = Arc::clone(&self.storage);
+        let peers_ = peers.clone();
+        let broadcast_task = async move {
+            let interval = if let Some(x) = self.config.broadcast_interval {
+                x
+            } else {
+                return Result::<(), Error>::Ok(());
+            };
+            loop {
+                let peers = peers_.read().await.to_vec();
```
Comment (lemonholic) [inline network/src/dms.rs:360]:
```suggestion
                let peers = peers_.read().await;
```
ditto

File: network/src/dms.rs (hunk 2 above)
Comment (junha1) [reply network/src/dms.rs:92]: RPC interface is for network communication. Anyhow::error can't be serialized into a network packet

File: network/src/dms.rs (hunk 1 above)
Comment (junha1) [reply network/src/dms.rs:194]: updated

File: network/src/dms.rs (hunk 3 above)
Comment (junha1) [reply network/src/dms.rs:245]: updated

File: network/src/dms.rs (hunk 5 above)
Comment (junha1) [reply network/src/dms.rs:346]: updated

File: network/src/dms.rs (hunk 6 above)
Comment (junha1) [reply network/src/dms.rs:360]: updated

File: network/src/dms.rs (hunk 2 above)
Comment (lemonholic) [reply network/src/dms.rs:92]: OK

File: network/src/dms.rs (hunk 3 above)
Comment (lemonholic) [reply network/src/dms.rs]: checked

**Review by lemonholic — APPROVED** [review]

**Pull request merged**